### ISSUE TITLE
NullReferenceException when message type is not found by assembly scan

### ISSUE THREAD

**Issue by slahn** [Particular/NServiceBus#4472]
As mentioned in #4471, when the NServiceBus assembly scanning does not find a message type,
we get a NullReferenceException:

    at NServiceBus.Pipeline`1.Invoke(TContext context) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Pipeline\Pipeline.cs:line     33
    at NServiceBus.MessageOperations.SendMessage(IBehaviorContext context, Type messageType, Object message, SendOptions options) in     C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\MessageOperations.cs:line 74
    at NServiceBus.MessageSession.Send(Object message, SendOptions options) in     C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\MessageSession.cs:line 15
    at NServiceBus.RunningEndpointInstance.Send(Object message, SendOptions options) in     C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Unicast\RunningEndpointInstance.cs:line 69
    at NServiceBusAssemblyScanCrash.Program.Main(String[] args) in     F:\Dokumenter\code\NServiceBusAssemblyScanCrash\NServiceBusAssemblyScanCrash\Program.cs:line 33

In our case, this happens because the message type is present in the app domain, but not in an assembly in the app domain base directory.

I'm not sure how much I can mess with the app domain in a unit test, so I made a minimal console program that reproduces the problem: https://gist.github.com/slahn/e270e8be28c27be08d14949e1de91e75

**Comment by timbussmann** [comment]
hey @slahn 

Thanks a lot for the repro 👍 
As far as I can tell, the problem is caused because we **only** scan assemblies in the base directory. We also rely on the nsb assembly to be included in the assembly scanning process, which doesn't happen if the nsb dll isn't in the application base directory. This causes the weird exceptions in the end.

The Assembly.Load call in your repro doesn't help because we don't scan loaded assemblies but we only scan assemblies in the directory.

A potential fix could be that we always scan the NServiceBus assembly. I'm not sure but I think the nsb assembly should always be loaded in the appdomain?
It might make more sense to scan all loaded or all referenced assemblies. The later might be tricky as we would have to get to the user's application assembly somehow?

Thoughts @Particular/nservicebus-maintainers ?

**Comment by timbussmann** [comment]
Hey @slahn 
I raised a PR in #4485 which should be able to resolve your issues by allowing AppDomain assemblies to be scanned. Would you be able to verify the changes fix your issues if I provide you a version containing the mentioned changes?

On a sidenote: I noticed in your sample that you're missing an `await` when sending messages from the endpoint. All endpoint operations require an await to work correctly and without race conditions. I assume this is just a mistake int he repro, but just to make sure that you keep this in mind for your production code.

**Comment by slahn** [comment]
I'll be able to build from your branch myself, most likely. But I won't be able to try this before next week.

Re: your sidenote; yes, I noticed this about an hour after I posted the issue, but since the repro failed consistently I didn't bother to update it. It's an issue in the repro only.

**Comment by timbussmann** [comment]
@slahn awesone, would love to hear back from you.
Please just make sure to actually enable the appdomain scanning using the following configuration when using the PR branch:
```
endpointConfiguration.AssemblyScanner(cfg => {
  cfg.ScanAppDomainAssemblies = true;
  // this might be required in case some assemblies are loaded into the appdomain and also found as dlls in the base folder
  cfg.ThrowExceptions = false;
});
```

**Comment by timbussmann** [comment]
@slahn have you had the chance to verify whether the proposed changes resolve your issues?

**Comment by timbussmann** [comment]
I'm closing this since we assume this should be resolved in 6.2. Please comment in case you are still facing this issue @slahn.
